xswap/hash_table.cpp

```diff
@@ -0,0 +1,46 @@
+#include <cstdlib>
+# include <cstring>
+#include "hash_table.h"
+
+EdgeHashTable::EdgeHashTable(int max_source, int max_target) {
+    int max_pair[2] = {max_source, max_target};
+    int size_of_hash = ((cantor_pair(max_pair) + 1) / CHAR_SIZE) + 1;
+    create_hash_table(size_of_hash);
+}
+
+bool EdgeHashTable::contains(int *edge) {
+    int edge_cantor = cantor_pair(edge);
+    return (bool)get_bit(hash_table[edge_cantor / CHAR_SIZE], edge_cantor % CHAR_SIZE);
+}
+
+void EdgeHashTable::add(int *edge) {
+    int edge_cantor = cantor_pair(edge);
+    set_bit_true(&hash_table[edge_cantor / CHAR_SIZE], edge_cantor % CHAR_SIZE);
+}
+
+void EdgeHashTable::remove(int *edge) {
+    int edge_cantor = cantor_pair(edge);
+    set_bit_false(&hash_table[edge_cantor / CHAR_SIZE], edge_cantor % CHAR_SIZE);
+}
+
+void EdgeHashTable::create_hash_table(int hash_size) {
+    hash_table = (char*)calloc(hash_size, sizeof(char));
```

Comment:
I guess you can use `CHAR_SIZE` here instead of `sizeof(char)` to squeeze a little more performance boost.